### PR TITLE
ApplicationTags custom tags form env vars.

### PR DESCRIPTION
**Pull request by laullon**
https://wavefront.atlassian.net/browse/MONIT-18302

Usage sample:
```
        ApplicationTags appTags = new ApplicationTags.Builder("my_app", "main")
                .tagsFromEnv("my_.*")
                .tagFromEnv("VERSION", "app_version")
                .build();
```

Sample app: 
* <https://github.com/laullon/wf-java-apptags-sample>
* <https://github.com/laullon/wf-java-apptags-sample/blob/master/my-app/src/main/java/com/mycompany/app/App.java#L20>

### CODE REVIEW

File: src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java [hunk 1]
```diff
@@ -82,7 +85,45 @@ public Builder shard(String shard) {
      * @return {@code this}
      */
     public Builder customTags(Map<String, String> customTags) {
-      this.customTags = customTags;
+      this.customTags.putAll(customTags);
+      return this;
+    }
+
+    /**
+     * Set additional custom tags from environment variables that matches the a
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java:93]: `Set additional custom tags from environment variables that match the given regex prefix`.

Can remove the part about this being optional. It's implict in the builder pattern.

File: src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java [hunk 2]
```diff
@@ -82,7 +85,45 @@ public Builder shard(String shard) {
...
+            if (!Strings.isNullOrEmpty(value)) {
+              this.customTags.put(var, value);
+            }
+        }
+      });
+      return this;
+    }
+    
+    /**
+     * Set additional a custom tag from environment variable.
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java:115]: `Set a custom tag from the given environment variable`. Ditto: remove the optional comment.

**Comment by vikramraman** [comment]
Left some minor doc comments. lgtm otherwise.

File: src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java [hunk 3]
```diff
@@ -82,7 +85,42 @@ public Builder shard(String shard) {
...
+      return this;
+    }
+
+    /**
+     * Set additional custom tags from environment variables that match the given regex.
+     *
+     * @param regex Regular expression.
+     * @return {@code this}
+     */
+    public Builder tagsFromEnv(String regex) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java:98]: Just realized that Application Tags does not have a unit test
Can you add a new Java class ApplicationTagsTest
https://github.com/wavefrontHQ/wavefront-sdk-java/tree/master/src/test/java/com/wavefront/sdk/common
and test this new functionality as well as current/old functionality in ApplicationTags?
Sorry for asking you to do this, but we should add this test. Can you do that?

Comment (laullon) [reply src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java:98]: sure

File: src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java [hunk 4]
```diff
@@ -82,7 +85,50 @@ public Builder shard(String shard) {
...
+        }
+      });
+      return this;
+    }
+    
+    /**
+     * just for unit tests
+     * @return 
+     */
+    Map<String,String> getenv() {
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java:116]: Can add a `@VisibleForTesting` annotation similar to this: https://github.com/wavefrontHQ/wavefront-sdk-java/blob/7fd1842fb4ef8c0670b5c6d09f87a8c49bb9a223/src/main/java/com/wavefront/sdk/common/clients/service/ReportingService.java#L88

File: src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java [hunk 5]
```diff
@@ -82,7 +85,50 @@ public Builder shard(String shard) {
...
+            if (!Strings.isNullOrEmpty(value)) {
+              this.customTags.put(var, value);
+            }
+        }
+      });
+      return this;
+    }
+    
+    /**
+     * just for unit tests
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/application/ApplicationTags.java:113]: Can remove the javadoc. Add the annotation as mentioned and the method is self explanatory.

**Pull request merged**